rpc/src/module/subscription.rs

```diff
@@ -0,0 +1,170 @@
+use ckb_logger::error;
+use ckb_notify::NotifyController;
+use crossbeam_channel::select;
+use jsonrpc_core::{futures::Future, Metadata, Result};
+use jsonrpc_derive::rpc;
+use jsonrpc_pubsub::{
+    typed::{Sink, Subscriber},
+    PubSubMetadata, Session, SubscriptionId,
+};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::sync::Arc;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    RwLock,
+};
+use std::thread;
+
+#[derive(Clone, Debug)]
+pub struct SubscriptionSession {
+    pub(crate) subscription_ids: Arc<RwLock<HashSet<SubscriptionId>>>,
+    pub(crate) session: Arc<Session>,
+}
+
+impl SubscriptionSession {
+    pub fn new(session: Session) -> Self {
+        Self {
+            subscription_ids: Arc::new(RwLock::new(HashSet::new())),
+            session: Arc::new(session),
+        }
+    }
+}
+
+impl Metadata for SubscriptionSession {}
+
+impl PubSubMetadata for SubscriptionSession {
+    fn session(&self) -> Option<Arc<Session>> {
+        Some(Arc::clone(&self.session))
+    }
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Hash)]
+#[serde(rename_all = "snake_case")]
+pub enum Topic {
+    NewTipHeader,
+    NewTipBlock,
+}
+
+#[allow(clippy::needless_return)]
+#[rpc(server)]
+pub trait SubscriptionRpc {
+    type Metadata;
+
+    #[pubsub(subscription = "subscribe", subscribe, name = "subscribe")]
+    fn subscribe(&self, meta: Self::Metadata, subscriber: Subscriber<String>, topic: Topic);
+
+    #[pubsub(subscription = "subscribe", unsubscribe, name = "unsubscribe")]
+    fn unsubscribe(&self, meta: Option<Self::Metadata>, id: SubscriptionId) -> Result<bool>;
+}
+
+type Subscribers = HashMap<SubscriptionId, Sink<String>>;
+
+#[derive(Default)]
+pub struct SubscriptionRpcImpl {
+    pub(crate) id_generator: AtomicUsize,
+    pub(crate) subscribers: Arc<RwLock<HashMap<Topic, Subscribers>>>,
+}
+
+impl SubscriptionRpc for SubscriptionRpcImpl {
+    type Metadata = Option<SubscriptionSession>;
+
+    fn subscribe(&self, meta: Self::Metadata, subscriber: Subscriber<String>, topic: Topic) {
```

Comment:
What if the subscriber makes a duplicated subscription to the same topic? It seems that it will receive the same notify-messages multiple times. So my question is, should we check the duplicated subscription?

Reply:
I don't have any preference about duplicate checking, however, it's not easy to implement the checking, because of the active subscriptions of session are private: https://github.com/paritytech/jsonrpc/blob/master/pubsub/src/subscription.rs#L18 , and we need to create a PR for jsonrpc crate first, then add some extra logic code.